A90-health-service-list-method.md

```diff
@@ -0,0 +1,82 @@
+# A90: Add List Method to gRPC Health Service
+----
+* **Author(s):** @marcoshuck
+* **Approver:** @markdroth
+* **Status:** In Review  
+* **Implemented in:** TBD
+* **Last updated:** 2025-02-22  
+* **Discussion at:** https://groups.google.com/g/grpc-io/c/tEI5G9sX0zc
+
+
+## Abstract
+
+This proposal introduces a new `List` RPC method for the Health service, allowing clients to retrieve the statuses of all monitored services or a filtered list of specific services. This feature simplifies integration with status-reporting dashboards and enhances observability for microservices.
+
+## Background
+
+The existing Health service provides basic health check functionality but lacks a mechanism to retrieve a comprehensive list of all monitored services and their statuses.
+
+This limitation makes it challenging for clients to aggregate health information across multiple services, particularly for use cases like publishing service statuses to dashboards such as [Cachet](https://cachethq.io/) or [Statuspage](https://www.atlassian.com/software/statuspage).
+
+It's important to keep in mind that the list of health services exposed by an application can change over the lifetime of the process.
+
+Kubernetes provides a similar capability with its `/readyz?verbose` endpoint, which lists the status of all components. This proposal aims to bring analogous functionality to the Health service, providing a unified view of service health.
+
+### Related Proposals
+
+N/A
+
+## Proposal
+
+Introduce a new `List` RPC method in the Health service with the following features:
+
+- Retrieve the health statuses of all monitored services or a filtered list of specified services.
+- Ensure the implementation is idempotent and side-effect free.
+- Provide a clear schema for the request and response to facilitate integration with external tools.
+
+### Proposed API Changes
+
+```protobuf
+message HealthListRequest {
+  repeated string services = 1; // Contains the name of the services that should be obtained. If empty, must return all.
+}
+
+message HealthListResponse {
+  map<string, HealthCheckResponse> statuses = 1; // Contains all the services and their respective status.
```

Comment:
Should there be any concern about the length of this in the result?  I know we don't expect it to be long.  But it seems at the very least, a generic gRPC-provided implementation of the service should have some kind of safeguard here.  E.g. restrict it to 100 entries, error if there are too many, etc.

I suppose we can always add such things later if it turns out to be a problem, but I thought I'd call it out now.

Reply:
We talked about this internally (although I think you may have been out that day).  In general, we expect the number of services to be fairly small here, so it doesn't seem worth adding the complexity of supporting pagination here.  If we ever run into a problem where the size is too large here, we can discuss adding something like that at that point.

Reply:
I'd be more than happy to help adding pagination if we ever face a problem like this.

Reply:
I think if I'm reviewing an implementation of the service, I'd expect us to have some kind of safeguard in there.  E.g. if there are more than 100 entries in the service, it would either return an error or truncate the response.

I don't feel too strongly about exactly what the behavior should be, but I do think there should be something in place, ideally specified or at least mentioned somewhere in this gRFC.

I disagree with the philosophy of "we don't think it will be a problem, so we'll just let it be unbounded until someone OOMs, and then we'll fix it".  I totally agree that it's not likely to be a problem for anyone, ever.  But adding a limiter now only takes 5 extra minutes just in case.

Reply:
If you are going to put in a limit, 1000 would be much better than 100.  Even though 100 seems like an unreasonably large number, someone going crazy with microservices might hit it.  If the concern is OOM, then even a million would be okay.

Reply:
A million starts to get questionable.  I'd be fine with a limit field and allowing it to go up to 10k or so.

I'd also be fine with hard-coding a limit of something like 10k for now to make it simple.  We can add a limit field later if we need it.

Like I said, I don't feel strongly about how we do it, just that we should have _something_ in place.  I'd _prefer_ it be something very lightweight and easy to implement, like a hard-coded 10k limit, but I won't fight if others prefer something more complicated.

Reply:
We met and discussed this and agreed that erroring with RESOURCE_EXHAUSED if there are more than 100 results sounds like the most reasonable.  We can add a limit field, etc in the future if the need ever arises, though nobody seems to think it will.